RUST_SECURE_CODING.md

```diff
@@ -0,0 +1,169 @@
+
+### Data Structures with Deterministic Internal Order
+
+Certain data structures, like HashMap and HashSet, do not guarantee a deterministic order for the elements stored within them. This lack of order can lead to problems in operations that require processing elements in a consistent sequence across multiple executions. In the Aptos blockchain, deterministic data structures help in achieving consensus, maintaining the integrity of the ledger, and ensuring that computations can be reliably reproduced across different nodes.
+
+Below is a list of deterministic data structures available in Rust. Please note, this list may not be exhaustive:
+
+- **BTreeMap:** maintains its elements in sorted order by their keys.
+- **BinaryHeap:** It maintains its elements in a heap order, which is a complete binary tree where each parent node is less than or equal to its child nodes.
+- **Vec**: It maintains its elements in the order in which they were inserted. ⚠️
+- **LinkedList:** It maintains its elements in the order in which they were inserted. ⚠️
+- **VecDeque:** It maintains its elements in the order in which they were inserted. ⚠️
+
+## Cryptography
+
+### No Custom Cryptographic Algorithm
+
+Use exclusively the cryptographic primitives exposed by the `aptos-crypto` crate.
+
+### Cryptographic Material Management
+
+Adhere strictly to established protocols for generating, storing, and managing cryptographic keys. This includes using secure random sources for key generation, ensuring keys are stored in protected environments, and implementing robust management practices to handle key lifecycle events like rotation and revocation [Key Management Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Key_Management_Cheat_Sheet.html).
+
+### Zeroing Sensitive Data
+
+Use [zeroize](https://docs.rs/zeroize/latest/zeroize/#) for zeroing memory containing sensitive data.
+
+## Misc
+
+### Forget and Memory Leaks
+
+Avoid using `std::mem::forget` in secure development, or any other function that leaks the memory.
+
+> Reference cycles can also cause memory leakage [[Rustbook: leak]](https://doc.rust-lang.org/book/ch15-06-reference-cycles.html?highlight=leak#reference-cycles-can-leak-memory).
+
+Most memory leaks result in general product reliability problems. If an attacker can intentionally trigger a memory leak, the attacker might be able to launch a denial-of-service attack (by crashing or hanging the program).
+
+### Fuzzing
+
+Aptos contains harnesses for fuzzing crash-prone code like deserializers, using [`libFuzzer`](https://llvm.org/docs/LibFuzzer.html) through [`cargo fuzz`](https://rust-fuzz.github.io/book/cargo-fuzz.html). For more examples, see the `testsuite/fuzzer` directory where find detailed README.md.
+
+## Appendix
+
+#### References
```

Comment:
this should probably be after `Conclusion` as `References` are commonly added after the content. possibly with a lower indent?